In [22]:
import numpy as np
import pandas as pd
from pathlib import Path
from sklearn.preprocessing import StandardScaler, MinMaxScaler, LabelEncoder
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression, Ridge, Lasso, ElasticNet
from sklearn.ensemble import RandomForestClassifier

In [3]:
train_df = pd.read_csv(Path('Resources/2019loans.csv'))
test_df = pd.read_csv(Path('Resources/2020Q1loans.csv'))

In [15]:
# Convert categorical data to numeric and separate target feature for training data
target = train_df.drop('loan_status', axis=1)
target = pd.get_dummies(target)
target.head()

Unnamed: 0   index  loan_amnt  int_rate  installment  annual_inc    dti  \
0       57107   57107    13375.0    0.1797       483.34    223000.0  29.99   
1      141451  141451    21000.0    0.1308       478.68    123000.0  11.26   
2      321143  321143    20000.0    0.1240       448.95    197000.0  11.28   
3       11778   11778     3000.0    0.1240       100.22     45000.0  18.08   
4      169382  169382    30000.0    0.1612      1056.49    133000.0  27.77   

   delinq_2yrs  inq_last_6mths  open_acc  ...  verification_status_Verified  \
0          0.0             0.0      15.0  ...                             0   
1          2.0             0.0      16.0  ...                             0   
2          0.0             0.0      12.0  ...                             0   
3          0.0             0.0      12.0  ...                             0   
4          0.0             2.0      13.0  ...                             0   

   pymnt_plan_n  initial_list_status_f  initial_list_status_w  \
0             1                      0                      1   
1             1                      0                      1   
2             1                      0                      1   
3             1                      0                      1   
4             1                      0                      1   

   application_type_Individual  application_type_Joint App  hardship_flag_N  \
0                            1                           0                1   
1                            1                           0                1   
2                            1                           0                1   
3                            1                           0                1   
4                            1                           0                1   

   hardship_flag_Y  debt_settlement_flag_N  debt_settlement_flag_Y  
0                0                       1                       0  
1                0                       1                       0  
2                0                       1                       0  
3                0                       1                       0  
4                0                       1                       0  

[5 rows x 94 columns]

In [16]:
# Convert categorical data to numeric and separate target feature for testing data
target = test_df.drop('loan_status', axis=1)
target = pd.get_dummies(target)
target.head()

Unnamed: 0  index  loan_amnt  int_rate  installment  annual_inc    dti  \
0       67991  67991    40000.0    0.0819       814.70    140000.0  19.75   
1       25429  25429     6000.0    0.1524       208.70     55000.0  11.52   
2       38496  38496     3600.0    0.1695       128.27     42000.0   6.74   
3       19667  19667    20000.0    0.1524       478.33    100000.0  12.13   
4       37505  37505     3600.0    0.1240       120.27     50000.0  16.08   

   delinq_2yrs  inq_last_6mths  open_acc  ...  \
0          0.0             1.0      18.0  ...   
1          2.0             0.0       8.0  ...   
2          0.0             0.0       6.0  ...   
3          0.0             2.0       7.0  ...   
4          0.0             3.0       6.0  ...   

   verification_status_Source Verified  verification_status_Verified  \
0                                    0                             0   
1                                    0                             0   
2                                    0                             0   
3                                    0                             0   
4                                    0                             0   

   pymnt_plan_n  initial_list_status_f  initial_list_status_w  \
0             1                      0                      1   
1             1                      0                      1   
2             1                      0                      1   
3             1                      0                      1   
4             1                      0                      1   

   application_type_Individual  application_type_Joint App  hardship_flag_N  \
0                            1                           0                1   
1                            1                           0                1   
2                            1                           0                1   
3                            1                           0                1   
4                            1                           0                1   

   hardship_flag_Y  debt_settlement_flag_N  
0                0                       1  
1                0                       1  
2                0                       1  
3                0                       1  
4                0                       1  

[5 rows x 93 columns]

In [13]:
# add missing dummy variables to testing set
add_value = LabelEncoder().fit_transform(test_df['loan_status'])
add_value

array([1, 1, 1, ..., 0, 0, 0])

In [26]:
X_train, X_test, y_train, y_test = train_test_split(target, add_value, random_state=1)

In [21]:
# Train the Logistic Regression model on the unscaled data and print the model score
model1 = LinearRegression().fit(target, add_value)
print(f'The model score is: {model1.score(target, add_value)}')

The model score is: 0.42057818119407897


In [23]:
# Train a Random Forest Classifier model and print the model score
model2 = RandomForestClassifier(random_state=1, n_estimators=100).fit(target, add_value)
print(f'The model score is: {model2.score(target, add_value)}')

The model score is: 1.0


In [24]:
# Scale the data
scaler = StandardScaler().fit(X_train)
X_train_scl = scaler.transform(X_train)
X_test_scl = scaler.transform(X_test)

In [31]:
# Train the Logistic Regression model on the scaled data and print the model score
model3 = LinearRegression().fit(X_train_scl, y_train)
print(f'The model score is: {model3.score(X_train_scl, y_train)}, test score is: {model3.score(X_test_scl, y_test)}')

The model score is: 0.42717131960103416, test score is: -2.0652654159698194e+27


In [32]:
# Train a Random Forest Classifier model on the scaled data and print the model score
model4 = RandomForestClassifier(random_state=1, n_estimators=100).fit(X_train_scl, y_train)
print(f'The train score is: {model4.score(X_train_scl, y_train)}, test score is: {model4.score(X_test_scl, y_test)}')

The train score is: 1.0, test score is: 0.8937074829931972


The Random Forest Classifier model is better than the Logistic Regression model based on the test score. The score went from -2.06 to 0.89. The score for the Random Forest Classifier model is closer to 1 than the Logistic Regression model.